### ISSUE TITLE
Automatic login token cleanup?

### ISSUE THREAD

**Issue by JeGr** [desec-io/desec-stack#976]
Hi, 

logged into my desec account and created a new token for another service to set up DNS entries. That's where I found I had around 30 disabled login tokens still in my account that weren't cleaned up which dated back to over a year. Manually cleaning them up I ran into the rate limit -.- so had to come back later to finish it.

Is there no auto-cleanup of auto-disabled login tokens or is the job not working correctly? I don't see why I'd need an auto-created login token even 2-4 weeks later so was quite surprised there was stuff over a year ago?

Cheers
\jens

**Comment by peterthomassen** [comment]
Sorry for the rate limit, that's actually a little unfortunate.

There is no auto-cleanup job. Tokens are deleted upon request (e.g., during logout), and just like other API tokens, login tokens otherwise hang around, and depending in circumstances (e.g., time) might become invalid.

The leftover tokens are a symptom of not logging out of the session. While you can say that's fine in your use case, we also think the leftover tokens are not a big problem.

That said, perhaps it would indeed make sense to have them cleaned up. The question is how to do this cheaply, as the validity calculation is currently not done in the database, but in the backend application. Perhaps an alternative solution would be to clean up tokens whenever the list is viewed, so that cleanup is on-demand only (cheap) but still transparent to the user. @nils-wisiol ?

**Comment by codiflow** [comment]
+1 for this

Currently I have around 97 invalid login tokens in my list and have to scroll to the end to see my valid ones (the ones I use for deployments).

Would make perfect sense if there would be a button in the token management backend which says "Delete invalid tokens" or something like that. I would also be happy with an option to select all visible entries in bulk with an option to delete them afterwards.

But currently the process is a bit – let's say **cumbersome** 😅 You have to click on the trash bin at the end of an entry, wait until the popup comes up and click (!) on "Delete" (Enter or another key combination does not work). So speeding this up is kind of impossible using those "hacky keyboard methods" 🤪 So currently I just leave them where they are as lifetime is precious 😉

**Comment by peterthomassen** [comment]
I guess it would suffice to clean up invalid tokens that used to be login tokens (i.e., don't clean up tokens that were manually created as API tokens, and later expired).

@nils-wisiol When do you think it would be best to do the cleanup? (On each `GET tokens/` operation? On each sign-in? Something else?)

**Comment by JeGr** [comment]
@peterthomassen If I may?
I'd suppose that on each sign-in would be enough for most users as it wouldn't even amass tokens then. If you'd get logged out due to timeouts so the token is there and disabled and you log back in, it gets cleaned up so there shouldn't be large amounts of inactive tokens anymore that would need housekeeping to remove but only the last one that got you logged out. But perhaps I'm missing some cases?

**Comment by peterthomassen** [comment]
I tend to agree. Probably `GET tokens/` should not trigger this behavior, as `GET` is not expected to have side effects.
